### PR TITLE
Fix #4465: Properly call default param getters for nested JS ctors

### LINKED ISSUE #4465: Default parameters in constructors of nested JS classes cause invalid IR
Repro:

```scala
package helloworld

import scala.scalajs.js
import js.annotation._

class B {
  class A(x: String = "") extends js.Object {  }
}

object HelloWorld {
  def main(args: Array[String]): Unit = {
    val b = new B
    new b.A()
  }
}
```

expected: links,
get:
```
Cannot access module for non-module helloworld.B$A$
[error]   exported to JavaScript with @JSExport
```

The problem is that the call-site and the definition site do not agree on whether the companion is a class or a module:

```scala
captures: superClass$: any, arg$outer: helloworld.B
js class helloworld.B$A extends scala.scalajs.js.Object (via superClass$) {
  val $outer: helloworld.B
  def "constructor"(...rest: any): any = {
    val prep0: java.lang.String = if ((rest[0] === (void 0))) {
      mod:helloworld.B$A$.$lessinit$greater$default$1;Ljava.lang.String()
    } else {
      rest[0].asInstanceOf[java.lang.String]
    };
    val outer: helloworld.B = arg$outer;
    val x: java.lang.String = prep0;
    super();
    if ((outer === null)) {
      throw mod:scala.scalajs.runtime.package$.unwrapJavaScriptException;Ljava.lang.Throwable;Ljava.lang.Object(null)
    } else {
      this.helloworld.B$A::$outer = outer
    };
    (void 0)
  }
  static def helloworld$B$A$$$outer;Lhelloworld.B$A;Lhelloworld.B{$outer}(this{this}: any): helloworld.B = {
    this.helloworld.B$A::$outer.asInstanceOf[helloworld.B]
  }
}
```

```scala
class helloworld.B$A$ extends java.lang.Object {
  def $lessinit$greater$default$1;Ljava.lang.String{<init>$default$1}(): java.lang.String = {
    ""
  }
  constructor def <init>;Lhelloworld.B;V(outer{$outer}: helloworld.B) {
    this.java.lang.Object::<init>;V()
  }
}
```

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala [hunk 1]
```diff
@@ -723,8 +728,17 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
...
-        if (sym.isClassConstructor || static) genLoadModule(trgSym)
-        else js.This()(encodeClassType(trgSym))
+        if (sym.isClassConstructor || static) {
+          assert(trgSym.isLifted == captures.nonEmpty,
+              "inconsistent lifted symbol / captures")
+
+          if (!trgSym.isLifted)
+            genLoadModule(trgSym)
+          else
+            genNew(trgSym, trgSym.primaryConstructor, captures)
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala:738]: I'm pretty sure this is incorrect. We could create several instances of the module per enclosing instance, if we use a `new` every time. We should use the *getter* method on the enclosing instance instead. If the enclosing instance is a JS class, that means accessing the property of its `JSName`; if it is a Scala class, it means calling the Scala method with its name and 0 argument.

Comment (sjrd) [reply compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala:738]: We should add a side-effect in the constructor of the companion object in the tests, and test that it is only applied once per enclosing instance.

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by gzm0** [comment]
Uh oh, this sounds more complicated. Shall we postpone this past 1.6.0?

**Comment by sjrd** [comment]
Yes, let's postpone it.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala [hunk 2]
```diff
@@ -1584,8 +1584,15 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
...
-    private def genJSClassCtorDispatch(sym: Symbol, allParams: List[Symbol],
+    private def genJSClassCtorDispatch(sym: Symbol, allParamSyms: List[Symbol],
         overloadNum: Int): (Exported, List[js.ParamDef]) = {
+      implicit val pos = sym.pos
+
+      val allParamsAndInfos = for {
+        (sym, info) <- allParamSyms.zip(jsParamInfos(sym))
+      } yield {
+        genVarRef(sym) -> info
+      }
```
[the remark this answers is not in the document]
Comment (sjrd) [reply compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:1595]: This comment still stands.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala [hunk 3]
```diff
@@ -719,8 +724,31 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
...
 
       val trgTree = {
-        if (sym.isClassConstructor || static) genLoadModule(trgSym)
-        else js.This()(encodeClassType(trgSym))
+        if (sym.isClassConstructor || static) {
+          assert(trgSym.isLifted == captures.nonEmpty,
+              "inconsistent lifted symbol / captures")
+
+          if (!trgSym.isLifted) {
+            assert(captures.isEmpty, "expected empty captures")
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala:732]: Isn't that assertion already covered by
```scala
assert(trgSym.isLifted == captures.nonEmpty,
```
a few lines above?

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala [hunk 4]
```diff
@@ -719,8 +724,31 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
...
+            genLoadModule(trgSym)
+          } else {
+            assert(captures.size == 1, "expected exactly one capture")
+
+            // Find the module accessor.
+            val outer = enteringPhase(currentRun.picklerPhase)(trgSym.owner)
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala:738]: Does `trgSym.originalOwner` work? It doesn't require time travel, IIRC.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala [hunk 5]
```diff
@@ -719,8 +724,31 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
...
+            val name = enteringPhase(currentRun.typerPhase)(trgSym.unexpandedName)
+
+            val modAccessor = outer.info.members.lookupModule(name)
+            val receiver = captures(0)
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala:742]:
```suggestion
            val receiver = captures.head
```
?

**Review by sjrd — CHANGES_REQUESTED** [review]
The fix basically LGTM now. I have a few minor comments.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (gzm0) [reply compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala]: I have also renamed `sym` to `ctorSym`. It was confusing me as well :P

**Review by sjrd — APPROVED** [review]

**Pull request merged**